Add alpha option to set background transparency

mystal opened the pull request:
Works on OS X, but not on Linux due to a glutin bug: https://github.com/tomaka/glutin/issues/639.

There are a few things I'd like to clean up/look into, but I figure I can put it out to start review. Things I'd like to do:
* Add a validation check to the alpha read from the config file to make sure it's within 0.0 and 1.0, inclusive.
* Not sure if `alpha` is the best config option name for this, so decide on the best name.
* Not sure if `alpha` values should be a float from 0.0 to 1.0, or integers or whatever.

Addresses issue #232.

impowski commented:
Neat.

jwilm commented:
This looks pretty good overall. Couple of thoughts:
- Maybe `background_opacity` would be a good config name
- Seems like it shouldn't be set on the text (foreground)
- Float value between 0 and 1 sounds great.
- Validation is certainly necessary. One option is to add a `struct Alpha(f32)` and implement a deserialize on it that will prevent it from being initialized outside of that range. We can maintain the invariant then that an `Alpha` value always lies in [0, 1.0].

mystal commented:
All great suggestions, thanks! Working on the changes.

By the way, if I don't set the alpha value on the foreground, I get black rectangular borders around glyphs. Presumably the font atlas doesn't have an alpha channel so when we pull glyphs out we bring their surrounding rectangle. I'm not terribly familiar with the font rasterizing code, any tips on where to look to fix this?

mystal commented:
Okay, I think I figured out how to fix the black bounding boxes around glyps (for the OS X font rendering path, still need to follow-up on Linux). I noticed that when we rendered a glyph, we were uploading it to the GPU with just RGB data. Now I am making sure to upload RGBA and am clearing the glyph's background with a transparent color. 

One wrinkle that's come up is that the fonts don't look as smooth as before. I'm sure it's something I've changed, maybe related to blending or anti-aliasing. Not sure yet. Any tips appreciated.

All pictures use the default `alacritty_macos` configuration with `use_thin_strokes: false` since I'm on a Macbook Air.
Before:
![before](https://cloud.githubusercontent.com/assets/74258/22177123/057c3e78-dfcd-11e6-864c-865feff11ae5.png)

After:
![after](https://cloud.githubusercontent.com/assets/74258/22177125/0fb9d6c0-dfcd-11e6-849e-df5e357e9b0b.png)

With transparency:
![transparent](https://cloud.githubusercontent.com/assets/74258/22177126/1492c058-dfcd-11e6-9051-decae653d9e7.png)

jwilm commented:
> One wrinkle that's come up is that the fonts don't look as smooth as before. I'm sure it's something I've changed, maybe I have to make some changes to blending or anti-aliasing. Not sure yet. Any tips appreciated.

Did you push the code for this? In any case, It may be that something is still assuming 3 bytes per pixel instead of 4.

mystal commented:
Just pushed it. Gonna start testing on my Arch Linux desktop to make sure the Freetype path works.

jwilm commented:
Oh, the new version might look bad since we use a per channel blending mode:

```rust
gl::BlendFunc(gl::SRC1_COLOR, gl::ONE_MINUS_SRC1_COLOR);
```

mystal commented:
I rebased and made a small change to the rendering. I rasterize the glyphs on an opaque black background, as alacritty currently does. But, I set alphaMask's alpha channel in the fragment shader to be the red channel of the glyph. Since glyphs are rasterized as white text, this should just be a luminance value, essentially.

Now the fonts look crisp as before with full opacity:
![screen shot 2017-03-01 at 18 23 12](https://cloud.githubusercontent.com/assets/74258/23490303/a47c26e8-feac-11e6-81bb-c6bfc407ddfd.png)

And with opacity set to 0.8, this is how it looks:
![screen shot 2017-03-01 at 18 24 08](https://cloud.githubusercontent.com/assets/74258/23490309/af19af62-feac-11e6-9cfd-56646bd57715.png)

Sadly, all is not well! When I switch windows and back, I get ghosts of glyphs appearing when they shouldn't!
![screen shot 2017-03-01 at 18 28 22](https://cloud.githubusercontent.com/assets/74258/23490342/f71dccf8-feac-11e6-9cd8-f4ae0f922fce.png)

I also notice that the window contents become a bit darker. Resizing the screen seems to fix it. No idea what's causing this yet... We clear the screen every frame, so I don't understand why these ghost glyphs would persist, even after I type more text. I don't see anything obvious, might have to dig into Glutin.

mystal commented:
After some tests, I've found that the same problem occurs in glutin's basic transparent example. However, the problem only occurs in the fork of glutin that alacritty uses. The latest version of glutin does not have it.

@jwilm Any idea how difficult it might be to update the glutin fork?

mystal commented:
Okay, so this weekend I fixed glutin's transparent window support on macOS and Linux: https://github.com/tomaka/glutin/pull/865

So once alacritty is using upstream glutin, I'll rebase this PR and it should _hopefully_ be good to go. I'll see how I can help out with #234.

jmacdonald commented:
@jwilm can we rebase the glutin fork to unblock this issue? Would love to see this merged! 😄

jmacdonald commented:
@mystal any chance you might implement background blurring in a follow up PR once this is merged in? 🙏 😄

PaulBGD commented:
Excited for this, any updates?

munyari commented:
@jwilm it does look like the upstream changes from glutin can be automatically merged into your fork (at least this is what the github interface is telling me).

jwilm commented:
Update: currently working on integrating non-fork version of Glutin/winit. Will be able to land this soon.

techhazard commented:
Really looking forward to this, do you have any updates?

jquick commented:
Would love to see this implemented!

jquick commented:
Was to excited for this so went ahead and did a rebase myself to help out @mystal 
https://github.com/jwilm/alacritty/pull/739

mystal commented:
Thanks for the enthusiasm, @jquick :) Seems like a lot of people really want this!

I have a free weekend coming up, so I'm going to rebase and try to implement this on Linux as well. Expect screenshot comparisons to iTerm2 and Tilix as well, which are the two terminals I use everyday right now.

mystal commented:
Okay, this is ready for a fresh review! I rebased, fixed some outstanding issues, and cleaned up the code:
* I made sure that backround colors are multiplied by the opacity for proper alpha blending.
* There's (currently) no need to introduce an alpha channel to the glyph cache, so I removed that.

I have a couple of quick screenshots showing side-by-side comparisons of transparency (at roughly 50% opacity) in Alacritty and other terminal emulators.

iTerm2
![iterm2_comparison](https://user-images.githubusercontent.com/74258/29491830-74083eec-851c-11e7-907c-2f399d5a9c5e.png)

Tilix
![tilix_comparison](https://user-images.githubusercontent.com/74258/29491831-7aee5a70-851c-11e7-98c7-6c31e06e0d92.png)

maximbaz commented:
Something is fishy with powerline symbols, at least on my Linux it looks like this with `background_opacity: 0.5` on a solid gray background:

![image](https://user-images.githubusercontent.com/1177900/29493653-b0f245d6-859a-11e7-8c6d-aabbdc133357.png)

maximbaz commented:
More examples, how neovim looks with and without opacity:

![image](https://user-images.githubusercontent.com/1177900/29493680-570cdff8-859b-11e7-9066-1f288e3b46eb.png)

maximbaz commented:
There are also a couple of issues when the alacritty window is being resized:

1. It flickers a lot showing the background image for a split second.
2. While in nvim, resizing back and forth cuts the text that was visible before.

Better see the gif:

![alacritty-opacity](https://user-images.githubusercontent.com/1177900/29494473-59f15c02-85ab-11e7-80e0-41accca2d2d3.gif)

maximbaz commented:
Once in a while besides cutting text out there also appears a gray or orange stripe on the right side of the terminal. I have orange color in the background image, so this is probably some side-effect of the flickering.

![image](https://user-images.githubusercontent.com/1177900/29494725-f1129542-85b0-11e7-9f33-1625f2cef9a2.png)

![image](https://user-images.githubusercontent.com/1177900/29494739-24b72ea8-85b1-11e7-8af0-16989867a113.png)

I'm using `compton` with the following [.compton.conf](https://github.com/maximbaz/dotfiles/blob/master/.compton.conf), I wonder if these flickering issues could be solved by using a different compton configuration, or they are truly coming from alacritty itself.

jwilm commented:
@mystal thanks for updating the PR! I'll prioritize landing this one since it's been such a long time in the making (my bad) and is current with master.

@maximbaz thanks for the great test drive report! The stripe on the side is totally my fault. When fixing the remaining BCE issues, it caused the stripe to show up when growing window. It's easy to fix and completely unrelated to this PR.

> While in nvim, resizing back and forth cuts the text that was visible before

This wasn't an issue without the transparency PR? Kind of surprising to get that behavior..

> Powerline arrow transparency

How does this work in other terminal emulators? I imagine like there's no way to say "powerline symbols should be transparent". I bet it's more like "colored backgrounds are never transparent", and then the opaque arrow continues to blend in.

> Flickering

Hard to say anything conclusive about this problem. I encountered an issue on i3wm where horizontally "split" terminals would lose their transparency and show garbage in the background. Also using compton.

@maximbaz would you mind testing the nvim resize without this PR?

@mystal re: powerline symbols -- I think we should investigate what other terminals do for colored backgrounds. I suspect they are always opaque.

mystal commented:
@maximbaz Thanks for the testing! I'll admit hadn't done much resize testing, so I'll do that and try to reproduce the bugs you're seeing.

@jwilm re: powerline symbols - I performed a quick test on iTerm2 and saw the same behavior, i.e. fully solid powerline symbols. Supposedly, there's an [option](https://github.com/gnachman/iTerm2/pull/277) to only make the main background color transparent, but I've yet to find it. I'll keep digging and also check terminals on Linux.

mystal commented:
re: resizing - with and without this PR, I'm seeing existing text get cut off when resizing to a smaller size and back.

Before resize:
![before](https://user-images.githubusercontent.com/74258/29496359-c9d6e1fc-8585-11e7-92e3-25322728fcf6.png)
After resize:
![after](https://user-images.githubusercontent.com/74258/29496360-cbd1583e-8585-11e7-92fa-7414f051982d.png)

This is on Linux with i3wm, but I'm seeing the same behavior on macOS.

maximbaz commented:
Powerline transparency in other terminals: no idea honestly, I don't have any other terminal configured with transparency at the moment. Although it is visible at the opacity level of `0.5`, at the level of `0.95` it is quite bearable, as the color difference is rather small 😃 

Resizing cut-off: I reported this issue because it didn't reproduce from my alacritty that is installed globally, but now I understand the reason :man_facepalming: - my default alacritty runs tmux automatically, and when using tmux there is no cut-off, tmux takes care of that. But when I built alacritty from branch, I didn't start tmux 😃 So yes, not an issue of this PR.

Stripe: good to know the issue is not caused by this PR 🙂

Flickering: it happens when decreasing the width of the terminal even on master, the difference is that on master only the part of terminal that got decreased flickers, but with opacity enabled the entire window flickers sometimes, not only the part that got decreased. In other words, it was flickering even before, but after this PR the flickering issue is much more visible. Not sure how to explain it better...

jwilm commented:
> re: resizing - with and without this PR, I'm seeing existing text get cut off when resizing to a smaller size and back.

This is expected when not using a full screen application that handles reflows, but the issue @maximbaz mentioned was specifically within neovim.

jwilm commented:
Given the feedback, I'm leaning towards merging this now and addressing the powerline issue in a separate PR. There are two approaches we can take there:

1. Non-default background colors are rendered fully opaque
2. Allow reducing opacity on a font-by-font basis. This is easy enough since it just means adding an alpha channel to glyphs and defaulting to 1.0.

maximbaz commented:
> This is expected when not using a full screen application that handles reflows, but the issue @maximbaz mentioned was specifically within neovim.

So to confirm, I think now it is the same issue, all comes down to the reflow handling. In tmux the issue is not reproducing, as tmux is taking care of reflow. Including neovim in my description apparently was unnecessary, I didn't notice it from the start :)

Agree on merging!

jwilm commented:
Thanks again @mystal for the PR, bringing it up-to-date, and your patience. Thanks @maximbaz for all the testing.
